Allow both integer and string request IDs in LSP

mprihoda opened the pull request:
Zed's LSP support expects request messages to have integer ID, Metals LSP uses string. According to specification, both is acceptable:

interface RequestMessage extends Message {

	/**
	 * The request id.
	 */
	id: integer | string;

...
This pull requests modifies the types and serialization/deserialization so that string IDs are accepted.

mprihoda committed:
Allow both integer and string request IDs in LSP.

The LSP specification allows both integer and string:
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#requestMessage

hearnadam commented on crates/lsp/src/lsp.rs:
```diff
@@ -876,7 +884,7 @@ impl LanguageServer {
             .map(|handlers| {
                 let executor = executor.clone();
                 handlers.insert(
-                    id,
+                    RequestId::Int(id),
```
I'm reviewing on mobile, so the context is a bit hard to see - will this require a follow up change for passing string ids?

mprihoda replied:
No, I think this should be ok. These are the IDs that are generated by Zed, always integer for now, and I do not see a reason to change it, as both integer and strings are allowed in spec.

SomeoneToIgnore commented on crates/lsp/src/lsp.rs:
```diff
@@ -1340,4 +1348,15 @@ mod tests {
+        let json = r#"{"jsonrpc":"2.0","id":"2","method":"workspace/configuration","params":{"items":[{"scopeUri":"file:///Users/mph/Devel/personal/hello-scala/","section":"metals"}]}}"#;
+        let notification = serde_json::from_str::<AnyNotification>(json);
+        assert!(
+            notification.is_ok(),
```
Let's rewrite this test a bit: do an `expect` on the `notification` and here, assert the the parsed id is actually 2 and it's a string.

I would add a test for serialization of the integers too, as this is something we use everywhere and it would be great to prove that it works (I've checked the branch and know that it works, but still: it's hard to regress things when there's a test).

Bonus tests: if you do not mind, let's also add a similar test for non-integer-like string + integer deserialization?

mprihoda replied:
Tests updated.

SomeoneToIgnore commented on crates/lsp/src/lsp.rs:
```diff
@@ -92,13 +92,21 @@ pub enum Subscription {
     },
 }
 
+/// ID is either string or number
```
This rustdoc does not make much sense, as repeats exactly what this `enum` declares.
I would propose to link the spec here instead, just like `see <a href="https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#requestMessage">the specification</a> for possible values`.

Thinking of this more, the integer id should be `i32` actually, not `usize`? 
The spec defines https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#uinteger for unsized ints and this is not our case.

If we want to make it right, we should change to that type instead?

mprihoda replied:
Updated the rust doc, and as a second commit changed the ID type to i32.

SomeoneToIgnore reviewed (commented):
Thank you for separating it out of the Scala PR, this seems quite a fundamental change indeed to have it mixed with the other ones.

I'm curious to know how Scala LSP is supposed to react to our integer IDs sent and mixed with the string ones?
I expect no issues as long as those are unique and there's no overlap with the server ones (that we do not ensure anyway now, so seems tolerable), but is it true?

mprihoda commented:
> I'm curious to know how Scala LSP is supposed to react to our integer IDs sent and mixed with the string ones?
> I expect no issues as long as those are unique and there's no overlap with the server ones (that we do not ensure anyway now, so seems tolerable), but is it true?

Metals works just fine with the integer IDs generated by Rust, it returns the integer back correctly.

mprihoda committed:
Use i32 for numeric LSP request ID.

Per specification, the ID is a 32bit integer.
https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#integer

SomeoneToIgnore reviewed (approved):
Beautiful, thank you for finally making us spec-compliant on that front.

I see an odd CI error and wondering if you could rebase on a fresh `main`? 
If that would not fix it, I'll just merge it since it seems unrelated, but let's be on a safe side.

SomeoneToIgnore commented:
Yes, here's the commit we want to pull into: https://github.com/zed-industries/zed/pull/7803

mprihoda commented:
Rebased, and improved the commit messages while I've been at it.
